keps/sig-instrumentation/4827-component-statusz/README.md

```diff
@@ -0,0 +1,406 @@
+* Set a precedent for presenting this basic but crucial component information in a consistent and accessible manner across the Kubernetes ecosystem
+
+### Non-Goals
+
+* Replace existing monitoring solutions: metrics, logs, traces
+* To provide a general-purpose extension mechanism for other components (e.g., container runtime, CNI, device plugins) to expose their own information through z-pages
+* Provide information about components that are not directly accessible due to network restrictions
+
+## Proposal
+
+We are proposing to add a new endpoint, /statusz on all core Kubernetes components which returns key details regarding a component's current state, including versioning, build information, and compatibility.
+
+
+### Risks and Mitigations
+
+1. **No sensitive data exposed**
+    
+    We will ensure that no sensitive data is exposed through statusz and that access to this endpoint is gated by using `system:monitoring group` ([ref](https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L73)).
+
+2. **Performance impact**
+    
+    We will take care to carefully consider the computational overhead associated with the statusz endpoint to avoid negatively impacting the server's overall performance.
+
+3. **Essential data only**
+    
+    We will prioritize inclusion of only essential diagnostic and monitoring data that aligns with the intended purpose of the z-page.
+
+4. **Premature Dependency on Unstable Format**
+
+    The alpha release will explicitly support plain text format only, making it clear that the output is not intended for parsing or automated monitoring. The feature will be secured behind a feature gate that is disabled by default, ensuring users opt-in consciously. Also, the endpoint will be implemented with versioned URLs to facilitate future transitions to structured formats without breaking existing integrations, once the usage patterns and requirements are better understood
+
+## Design Details
+
+### Data Format and versioning
+
+Initially, the statusz page will exclusively support a plain text format for responses. We will implement these endpoints using versioned URLs, like /statusz/v1, to ensure future compatibility. This versioning strategy allows us to seamlessly introduce structured data formats (e.g., JSON) in the future, through a distinct endpoint such as /statusz/v2, without disrupting existing implementations.
+
+### Authz and authn
+
+Access to the endpoint will be limited to members of the `system:monitoring group` ([ref](https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L73)), which is how paths like /healthz, /livez, /readyz are restricted today. 
+
+### Endpoint Response Format
+
+#### Data format : text
+
+#### Request
+* Method: **GET** 
+* Endpoint: **/statusz/v1**
```

Comment:
why is this one versioned? could you document the reasoning of this being versioned.

statusz is usually not versioned.

Reply:
I have explained the rationale behind having versioned endpoints under "Data Format and versioning", does that help?

Reply:
I think I would use headers or params for this, unless we want to treat it like the kube-API

Reply:
The idea was for statusz to be backwards compatible, in case when we change the response schema down the line. Versioned URLs would provide us with the option to be able to make those changes without breaking existing users.

Reply:
That said, I'm flexible on the specific implementation, whether it's through the URL path, headers, or query parameters.  My main goal is to ensure we have versioning support for this endpoint.

Reply:
I have now updated the KEP to specify support for versioning through url parameters instead of embedding it in the url path. It will help with differentiating this endpoint from kube-API and it also aligns with Google's internal guidance for z-pages.

Reply:
Based on [this](https://github.com/kubernetes/enhancements/pull/4830#discussion_r1792676964) comment thread, removed the version url param from the spec.